### PR TITLE
tracing: Allow setting some optional segment fields in X-Ray traces

### PR DESCRIPTION
**Pull request by lavignes**
Signed-off-by: Scott LaVigne <lavignes@amazon.com>

Commit Message:
In the closed-source implementation of the AWS X-Ray tracer extension
there was additional span metadata that was hard-coded for the AWS App Mesh
service. This was expectedly removed from the implementation that was
upstreamed but there was a miss in adding an API to re-add this metadata
to spans.

The extension emits traces that are technically functional, but some
helpful metadata is missing: aws/aws-app-mesh-roadmap#246

Additional Description: N/A
Risk Level: LOW (this is opt-in)
Testing: Updated unit tests.
Docs Changes: Added docs for new api fields.

### CODE REVIEW

**Comment by lavignes** [comment]
Woops. looks like something in the commit got messed up. Will update the PR :)

**Comment by lavignes** [comment]
Friendly ping for you to take a look at this @mattklein123 
Sorry for bugging you!

**Comment by lavignes** [comment]
ok merged from upstream

File: api/envoy/config/trace/v3/xray.proto [hunk 1]
```diff
@@ -35,4 +46,9 @@ message XRayConfig {
   // `X-Ray SDK documentation
   // <https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-go-configuration.html#xray-sdk-go-configuration-sampling>`_
   core.v3.DataSource sampling_rule_manifest = 3;
+
+  // Optional custom fields to be added to each trace segment.
+  // see: `X-Ray Segment Document documentation
+  // <https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html>`__
+  SegmentFields segment_fields = 4;
```
Comment (mattklein123) [inline api/envoy/config/trace/v3/xray.proto:53]: qq: should this be repeated? I see that the segment fields has an origin. What if you have different origins? I don't know the protocol but figured I would ask.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -91,6 +91,7 @@ New Features
 * tap: added :ref:`generic body matcher<envoy_v3_api_msg_config.tap.v3.HttpGenericBodyMatch>` to scan http requests and responses for text or hex patterns.
 * tcp: switched the TCP connection pool to the new "shared" connection pool, sharing a common code base with HTTP and HTTP/2. Any unexpected behavioral changes can be temporarily reverted by setting `envoy.reloadable_features.new_tcp_connection_pool` to false.
 * tcp_proxy: allow earlier network filters to set metadataMatchCriteria on the connection StreamInfo to influence load balancing.
+* tracing: added ability to set some `optional segment fields<envoy_v3_api_field_config.trace.v3.XRayConfig.segment_fields>` in the AWS  X-Ray tracer.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:94]: This is not a proper ref link. Please use `:ref:`. You can render the docs via `docs/build.sh` if you want to check it.

Comment (lavignes) [reply docs/root/version_history/current.rst:94]: My bad. Ill fix it.

File: api/envoy/config/trace/v3/xray.proto (hunk 1 above)
Comment (lavignes) [reply api/envoy/config/trace/v3/xray.proto:53]: I though about just making this entire field a struct and pass it through, but I wanted to try and match the x-ray schema 1:1. I also ran this by our X-Ray team and they signed off on this approach.

There is only ever one origin in the sense that it is the origin of the segment, but the entire trace will have multiple origins across its lifetime.

Comment (mattklein123) [reply api/envoy/config/trace/v3/xray.proto:53]: :+1:

**Pull request merged**